hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrngSystemPrecompiledContract.java

```diff
@@ -0,0 +1,291 @@
+				createSuccessfulChildRecord(randomNum, frame, input) :
+				createUnsuccessfulChildRecord(result, frame);
+
+		final var parentUpdater = updater.parentUpdater();
+		if (parentUpdater.isPresent()) {
+			final var parent = (AbstractLedgerWorldUpdater) parentUpdater.get();
+			parent.manageInProgressRecord(recordsHistorian, childRecord, body(randomNum, input));
+		} else {
+			throw new InvalidTransactionException("PRNG precompile frame had no parent updater", FAIL_INVALID);
+		}
+
+		return result.getLeft();
+	}
+
+	Pair<PrecompileContractResult, Exception> computePrngResult(
+			final long gasNeeded,
+			final Bytes input,
+			final MessageFrame frame) {
+		try {
+			validateTrue(frame.getRemainingGas() >= gasNeeded, INSUFFICIENT_GAS);
+			final var randomNum = generatePseudoRandomData(input);
+			return Pair.of(PrecompiledContract.PrecompileContractResult.success(randomNum), null);
+		} catch (InvalidTransactionException | IllegalArgumentException e) {
+			return Pair.of(PrecompiledContract.PrecompileContractResult.halt(null,
+					Optional.ofNullable(ExceptionalHaltReason.INVALID_OPERATION)), e);
+		} catch (Exception e) {
+			log.warn("Internal precompile failure", e);
+			return Pair.of(PrecompiledContract.PrecompileContractResult.halt(null,
+					Optional.ofNullable(ExceptionalHaltReason.INVALID_OPERATION)), e);
+		}
+	}
+
+	Bytes generatePseudoRandomData(final Bytes input) {
+		final var selector = input.getInt(0);
+		return switch (selector) {
+			case PSEUDORANDOM_SEED_GENERATOR_SELECTOR -> random256BitGenerator();
+			case PSEUDORANDOM_NUM_IN_RANGE_GENERATOR_SELECTOR -> randomNumberGeneratorInRange(input);
+			default -> null;
+		};
+	}
+
+	private Bytes random256BitGenerator() {
+		final var hashBytes = prngLogic.getNMinus3RunningHashBytes();
+		if (isEmptyOrNull(hashBytes)) {
+			return null;
+		}
+		return Bytes.wrap(hashBytes, 0, 32);
+	}
+
+	private Bytes randomNumberGeneratorInRange(final Bytes input) {
+		final var range = rangeValueFrom(input);
+
+		validateTrue(range >= 0, INVALID_PRNG_RANGE);
+
+		final var hashBytes = prngLogic.getNMinus3RunningHashBytes();
+		if (isEmptyOrNull(hashBytes)) {
+			return null;
+		}
+
+		final var randomNum = prngLogic.randomNumFromBytes(hashBytes, range);
+		return padded(randomNum);
+	}
+
+	private boolean isEmptyOrNull(final byte[] hashBytes) {
+		return hashBytes == null || hashBytes.length == 0;
+	}
+
+	long calculateGas(final Instant now) {
+		final var feesInTinyCents = pricingUtils.getCanonicalPriceInTinyCents(PRNG);
+		final var currentGasPriceInTinyCents = livePricesSource.currentGasPriceInTinycents(now,
+				HederaFunctionality.ContractCall);
+		return feesInTinyCents / currentGasPriceInTinyCents;
+	}
+
+	ExpirableTxnRecord.Builder createUnsuccessfulChildRecord(
+			final Pair<PrecompileContractResult, Exception> result, final MessageFrame frame) {
+		var response = FAIL_INVALID;
+		if (result.getRight() instanceof InvalidTransactionException) {
```

Comment:
Could we refactor just a bit and return `Pair<PrecompileContractResult, ResponseCodeEnum>` from `computePrngResult()`---where a thrown `IllegalArgumentException` returns `INVALID_PRNG_RANGE` as the right part of the pair?

As is, we will return `FAIL_INVALID` for an `IllegalArgumentException`.

Reply:
Fixed. Thanks !